src/palace/manager/marc/uploader.py

```diff
@@ -109,7 +109,10 @@ def complete(self) -> set[str]:
                     # Upload the last chunk if the buffer is not empty, the final part has no
                     # minimum size requirement.
                     upload_part = self.storage_service.multipart_upload(
-                        key, upload.upload_id, len(upload.parts), upload.buffer.encode()
+                        key,
+                        upload.upload_id,
+                        len(upload.parts) + 1,
```

Comment:
I simplified the logic here so its easier to follow, and the part logic with offsets is all encapsulated in the same place. 

The simplification comes at the expense of efficiency, since it will require a couple more calls to redis and another call to s3, but that inefficiency probably doesn't make much a difference here and its much more readable.